### DISCUSSION TITLE
Implicit Functions: Lathe

### DISCUSSION THREAD

**Discussion by andreasrauch** [AMReX-Codes/amrex#3256] [Q&A] [answered]
Hello,

I want to construct a cone using implicit functions. My thinking was to create a triangle and then rotate it with lathe. But, I'm unclear on how lathe works and am getting unexpected results.

Does lathe take whatever object is passed to it and simply rotate around the z-axis? Are there constraints on whether the object needs to centered on the origin? Can lathe be used to rotate around a different axis?

The amrex website says: lathe: Creates a surface of revolution by rotating a 2D object around an axis.
But how can I create a 2D object when running AMReX in 3 space dimensions?

   Geometry setup:
- 3D prism created: protruded in x-direction with y-z plane cross section as an isosceles triangle:   
        triangle 'tip'  <|    triangle 'base'     
- Tip points in the negative z-direction, base extends in y direction and shape is protruded across full x-length.
- The eb geometry setup code: [eb_geom.txt](https://github.com/AMReX-Codes/amrex/files/11224481/eb_geom.txt)

This is the geometry without lathe applied:
![vfrac_nolathe1](https://user-images.githubusercontent.com/116833831/231828588-94fd037a-7c11-4806-9cf4-1462f4205931.png)
![vfrac_nolathe2](https://user-images.githubusercontent.com/116833831/231828593-36300a52-5a96-4be7-89e4-016dc9d00256.png)

From tests I've done:
triangle tip on origin + lathe = all fluid
triangle base on origin + lathe = all fluid
triangle centered on origin + lathe = 3D object rotated around z-axis

This is the geometry when applying lathe (to the prism centered on the origin):
![vfrac_lathe](https://user-images.githubusercontent.com/116833831/231828586-dca16157-d3ac-4424-9011-ac0eaa714bc5.png)

Thanks,

Andreas

**Comment by nataraj2** [comment, marked as answer]
Try the following and let us know if it works or not.

In the inputs, 

```
geometry.prob_lo     =  -0.025 -0.05 -0.05
geometry.prob_hi     =   0.025  0.05  0.05

amr.n_cell           =   32 64 64
amr.max_level       = 3 
````

and use the following to create the cone

```.  
        EB2::SplineIF TriForCone;
        std::vector<amrex::RealVect> pts;
        amrex::RealVect p;

	p = amrex::RealVect(AMREX_D_DECL(0.0,0.0,0.0));
        pts.push_back(p);
	p = amrex::RealVect(AMREX_D_DECL(-0.577,1.0,0.0));
        pts.push_back(p);

        TriForCone.addLineElement(pts);

        pts.clear();
        auto Cone = EB2::lathe(TriForCone);

	auto ConeTranslate = EB2::translate(Cone,{0.0,0.0,0.0});
	RealArray point;
        RealArray normal;

        point[0]=0.0;point[1]=0.0;point[2]=-0.25;
	normal[0]=0.0;normal[1]=0.0;normal[2]=1.0;

	EB2::PlaneIF pf(point, normal);
        
        EB2::CylinderIF cf(0.01, 0.034641, 2, {0.0,0.0,0.0}, false);

        auto polys = EB2::makeIntersection(ConeTranslate,cf);
        auto poly_translate = EB2::translate(polys,{0.0,0.0,-0.02});

        auto gshop = EB2::makeShop(poly_translate);
        EB2::Build(gshop, geom, max_coarsening_level, max_coarsening_level, 4, false);
```

**Reply by andreasrauch** [reply]
This worked, thank you

**Comment by rmrsk** [comment]
No need to go through all of that trouble. Here's a simpler way:

    Real operator()(AMREX_D_DECL(Real x, Real y, Real z)) const noexcept {
      // Opening angle and height
      constexpr Real theta  = 30.;
      constexpr Real height = 1.0;

      constexpr Real C1 = cos(theta * M_PI / 180.);
      constexpr Real C2 = sin(theta * M_PI / 180.);
  
      return -std::max(C1 * sqrt(x * x + z * z) + C2 * y, -height - y);
    }
